services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,630 @@
+                            ));
+                        });
+                    });
+                })
+                .recover(errorVersionEqualsORNotSet -> {
+                    TracingHelper.logError(span, String.format("Tenant [%s] not found.", tenantId));
+                    return Future.<OperationResult<Void>>succeededFuture(OperationResult.<Void>empty(HttpURLConnection.HTTP_NOT_FOUND));
+                });
+
+
+    }
+
+    /**
+     * Deletes a present tenant.
+     * <p>
+     * Tenant modification have to be enabled in {@link MongoDbBasedTenantsConfigProperties} or exits with {@code HttpURLConnection.HTTP_FORBIDDEN}.
+     * </p>
+     * <p>
+     * Provided version {@code resourceVersion} have to be empty or equal to the version of present tenant.
+     * </p>
+     *
+     * @param tenantId        The identifier of the tenant.
+     * @param resourceVersion The identifier of the resource version to delete.
+     * @param span            The active OpenTracing span for this operation. It is not to be closed in this method!
+     *                        An implementation should log (error) events on this span and it may set tags and use this span as the
+     *                        parent for any spans created in this method.
+     * @return Result of deletion operation: On success {@code HttpURLConnection.HTTP_NO_CONTENT}, http error codes else. {@code HttpURLConnection.HTTP_FORBIDDEN} if modification is disabled.
+     * @throws NullPointerException if {@code tenantId} or resourceVersion are {@code null}.
+     */
+    @Override
+    public Future<Result<Void>> deleteTenant(final String tenantId, final Optional<String> resourceVersion, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(resourceVersion);
+
+        return executeIfModificationEnabled(
+                () -> processDeleteTenant(tenantId, resourceVersion, span),
+                span
+        );
+    }
+
+    /**
+     * Deletes a present tenant.
+     * <p>
+     * Provided version {@code resourceVersion} have to be empty or equal to the version of present tenant.
+     * </p>
+     *
+     * @param tenantId        The identifier of the tenant.
+     * @param resourceVersion The identifier of the resource version to delete.
+     * @param span            The active OpenTracing span for this operation. It is not to be closed in this method!
+     *                        An implementation should log (error) events on this span and it may set tags and use this span as the
+     *                        parent for any spans created in this method.
+     * @return Result of deletion operation: On success {@code HttpURLConnection.HTTP_NO_CONTENT}, http error codes else.
+     * @throws NullPointerException if {@code tenantId} or resourceVersion are {@code null}.
+     */
+    private Future<Result<Void>> processDeleteTenant(final String tenantId, final Optional<String> resourceVersion, final Span span) {
```

Comment:
Please follow the hono `javadoc` style and you could use `TenantManagementService` as a reference.